src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
+            .filter { score.score(it) >= 0 && !isDirect(it) }
+
+        if (meshPeers.size < coreParams.D) {
+            val addFromFanout = fanoutPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size)
+            val addFromOthers = otherPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size - addFromFanout.size)
 
-            meshPeers += (addFromFanout + addFromOthers)
             (addFromFanout + addFromOthers).forEach {
                 graft(it, topic)
             }
+            fanout -= topic
+            lastPublished -= topic
         }
     }
 
-    override fun unsubscribe(topic: String) {
+    override fun unsubscribe(topic: Topic) {
         super.unsubscribe(topic)
-        mesh.remove(topic)?.forEach { prune(it, topic) }
+        mesh[topic]?.forEach { prune(it, topic) }
     }
 
-    private fun heartBeat(time: Long) {
+    private fun heartBeat() {
+        heartbeatsCount++
+        iAsked.clear()
+        peerIHave.clear()
+
+        val staleIWantTime = this.curTime() - params.iWantFollowupTime.toMillis()
+        iWantRequests.entries.removeIf { (key, time) ->
+            (time < staleIWantTime)
+                .whenTrue { notifyRouterMisbehavior(key.first, 1) }
+        }
+
         try {
             mesh.entries.forEach { (topic, peers) ->
-                if (peers.size < DLow) {
-                    (getTopicPeers(topic) - peers).shuffled(random).take(D - peers.size).forEach { newPeer ->
-                        peers += newPeer
-                        graft(newPeer, topic)
-                    }
-                } else if (peers.size > DHigh) {
-                    peers.shuffled(random).take(peers.size - D).forEach { dropPeer ->
-                        peers -= dropPeer
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
+                peers.filter { score.score(it) < 0 }
+                    .forEach { prune(it, topic) }
+
+                if (peers.size < coreParams.DLow) {
+                    // need more mesh peers
+                    (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= 0 && !isDirect(it) && !isBackOff(it, topic) }
+                        .shuffled(random)
+                        .take(coreParams.D - peers.size)
+                        .forEach { graft(it, topic) }
+                } else if (peers.size > coreParams.DHigh) {
+                    // too many mesh peers
+                    val sortedPeers = peers
+                        .shuffled(random)
+                        .sortedBy { score.score(it) }
+                        .reversed()
+
+                    val bestDPeers = sortedPeers.take(coreParams.DScore)
+                    val restPeers = sortedPeers.drop(coreParams.DScore).shuffled(random)
+                    val outboundCount = bestDPeers.count { it.isOutbound() }
```

Comment:
I think this should be:
```suggestion
                    val outboundCount = (bestDPeers + restPeers).count { it.isOutbound() }
```
As the spec says:
> **if the scoring plus random selection** does not result in enough outbound connections, then we replace the random and lower scoring peers in the selection with outboud connection peers.